lib/app_funcs.sh

```diff
@@ -148,33 +139,47 @@ function pre_compile_hook() {
   cd - > /dev/null
 }
 
-function write_profile_d_script() {
-  output_section "Creating .profile.d with env vars"
-  mkdir -p $build_path/.profile.d
+function export_var() {
+  local VAR_NAME=$1
+  local VAR_VALUE=$2
+
+  echo "export ${VAR_NAME}=${VAR_VALUE}"
+}
 
-  local export_line="export PATH=\$HOME/.platform_tools:\$HOME/.platform_tools/erlang/bin:\$HOME/.platform_tools/elixir/bin:\$PATH
-                     export LC_CTYPE=en_US.utf8"
+function export_default_var() {
+  local VAR_NAME=$1
+  local DEFAULT_VALUE=$2
 
-  # Only write MIX_ENV to profile if the application did not set MIX_ENV
-  if [ ! -f $env_path/MIX_ENV ]; then
-    export_line="${export_line}
-                 export MIX_ENV=${MIX_ENV}"
+  if [ ! -f "${env_path}/${VAR_NAME}" ]; then
+    export_var "${VAR_NAME}" "${DEFAULT_VALUE}"
   fi
+}
 
-  echo $export_line >> $build_path/.profile.d/elixir_buildpack_paths.sh
+function echo_profile_env_vars() {
+  local buildpack_bin="$(runtime_platform_tools_path)"
+  buildpack_bin="$(runtime_erlang_path)/bin:${buildpack_bin}"
+  buildpack_bin="$(runtime_elixir_path)/bin:${buildpack_bin}"
+
+
+  export_var "PATH" "${buildpack_bin}:\$PATH"
+  export_default_var "LC_CTYPE" "en_US.utf8"
+
+  # Only write MIX_* to profile if the application did not set MIX_*
+  export_default_var "MIX_ENV" "${MIX_ENV}"
+  export_default_var "MIX_HOME" "$(runtime_mix_home_path)"
+  export_default_var "HEX_HOME" "$(runtime_hex_home_path)"
 }
 
-function write_export() {
-  output_section "Writing export for multi-buildpack support"
+function write_profile_d_script() {
+  output_section "Creating .profile.d with env vars"
+  mkdir -p $build_path/.profile.d
+  local profile_path="${build_path}/.profile.d/elixir_buildpack_paths.sh"
 
-  local export_line="export PATH=$(platform_tools_path):$(erlang_path)/bin:$(elixir_path)/bin:$PATH
-                     export LC_CTYPE=en_US.utf8"
+  echo_profile_env_vars >> $profile_path
+}
 
-  # Only write MIX_ENV to export if the application did not set MIX_ENV
-  if [ ! -f $env_path/MIX_ENV ]; then
-    export_line="${export_line}
-                 export MIX_ENV=${MIX_ENV}"
-  fi
+function write_export() {
+  output_section "Writing export for multi-buildpack support"
 
-  echo $export_line > $build_pack_path/export
+  echo_profile_env_vars >> "${build_pack_path}/export"
```

Comment:
If I'm following this correctly, it seems like the export file will contain runtime paths in `MIX_HOME` and `HEX_HOME`, rather than build time paths. When the export file is sourced prior to later buildpacks running, it's still build time, so the slug won't yet be extracted to `/app`. This might work fine since I see there are also copying steps that copy files to `/app` during the build, so `/app` may still contain what you need when later buildpacks run (`/app` is not purged between buildpacks) - but wanted to mention in case.

Reply:
Good catch; I've updated the mix install and erlang install functions to copy if the build time path and the runtime path aren't the same.

Reply:
@fivetanley What would you say are the pros/cons of the copying approach vs making sure the `export` file has the correct build-time `MIX_HOME` and `HEX_HOME` values? (Currently the `export` file has the run-time `/app` values).

With the copying approach, if a later buildpack eg installs additional dependencies, won't those end up in the `/app` copy, and not in build_dir, and so be missing at runtime?

Reply:
> vs making sure the `export` file has the correct build-time `MIX_HOME` and `HEX_HOME` values

To be clear, by this approach I mean copying `echo_profile_env_vars` and having another version that outputs paths that use `$BUILD_DIR` for use with the `export` file, rather than trying to use the output of `echo_profile_env_vars` for both the `profile.d` script and the `export` file.

Reply:
Ah, ok, I think I understand now. Can you review the updated diff (this convo is on the outdated diff) for this function?

Reply:
@fivetanley The latest version looks great - thank you! :-)